### PR TITLE
[bugfix] fix colorscale bug

### LINKED ISSUE #165: Marker color cannot be set using value from field
It is not possible to use a numerical field and a color scale from the dropdown to define the marker color. The marker color is instead set to a default value(`"#8ebad9"`), and the color scale in the plot stays empty.

See the attached GIF:

### PR DESCRIPTION
**Pull request by ghtmtt**
fix #165

### CODE REVIEW

**Comment by ghtmtt** [comment]
cc @nyalldawson 

A starting boilerplate....

still a little bit weird the output, but with this PR the plotly scale color approach is gone in favor of the QGIS one. We can use the assistant or also some nice expression (`CASE WHEN......`):

With expression method:

immagine


immagine

With color assistant method

immagine

**TODO**:

* with a simple field used as input it does not work as expected -> only single color
* the legend output is weird, don't know why
* ticks of the colorscale are missing (hided for now). We should retrieve the evaluation of the expression to get real values to put in the legend

File: DataPlotly/core/plot_factory.py [hunk 1]
```diff
@@ -278,6 +279,14 @@ def add_source_field_or_expression(field_or_expression):
             self.settings.data_defined_marker_sizes = marker_sizes
         if colors:
             self.settings.data_defined_colors = colors
+            try:
+                total = 1.0 / (len(colors) - 1)
+            except ZeroDivisionError:
+                total = 1.0
+            for current, color in enumerate(colors):
```
Comment (nyalldawson) [inline DataPlotly/core/plot_factory.py]: You'd need to do this from the source color ramp itself, not from the evaluated colors (which may be in any order). You should be able to get this by:

1. Check if a color ramp transformer is set on the property. If it's not, then we probably don't want a legend anyway OR we'd need a totally different approach for the legend (e.g. if categorised colors are being used)

2. Loop through the colours in that ramp, and convert those and their corresponding positions to the plotly ramp format.

3. From the property's transformer you can also get the min and max value range, so you'd scale the numeric values in the plotly ramp to fit between these ranges. That should give the correct values in the legend.

**Comment by ghtmtt** [comment]
With this last approach we are closer.

If the same values of the color (3rd variable) are used in the color scale legend, then plotly interpolates the colors in a  weird way. The solution seems to be to pass a lower amount of colors (1st, 25th, 50th, 75th and last element of the colors) to have a nicer view. @nyalldawson what do you think? If we decide to use this approach then we can drop the plotly color scale method with color scale names and use the better approach of QGIS with the assistant.

immagine

**Comment by nyalldawson** [comment]
@ghtmtt I've corrected the logic here, but something odd is happening still -- to me the color scale passed to plotly looks correct (eg. ``[0, '#ff0000'], [0.25, '#00ff00'] ... ]`, but the legend is always showing empty. Can you see what's missing?

File: DataPlotly/core/plot_factory.py [hunk 2]
```diff
@@ -270,6 +272,17 @@ def add_source_field_or_expression(field_or_expression):
                                                                               context, default_value)
                 stroke_colors.append(value.name())
 
+        # build up color scale for legend (if possible!)
+        if isinstance(self.settings.data_defined_properties.property(PlotSettings.PROPERTY_COLOR).transformer(),
```
Comment (ghtmtt) [inline DataPlotly/core/plot_factory.py]: I'm pretty sure the reason is that the legend wants as mandatory the first value with index 0 and the last with index 1. If the last value is not set as 1 index then nothing is rendered.

Comment (nyalldawson) [reply DataPlotly/core/plot_factory.py]: I've checked the contents of the array though, and it looks fine to me ...

Comment (ghtmtt) [reply DataPlotly/core/plot_factory.py]: Actually the current array contains a list of hex color in the `colors` parameter while it should contain a list of numbers, then the color_scale will work. We need to grab the field (or expression) from the assistant dialog, evaluate it and get a list of *real* values.

Comment (nyalldawson) [reply DataPlotly/core/plot_factory.py]: Oh, so you can't mix and match the two? Hmmmm, this gets incredibly complex now....

Comment (ghtmtt) [reply DataPlotly/core/plot_factory.py]: Seems not. Having real numbers for the color and the [index, hex] for the colors scale seems the best way to achieve it. 

Is it really getting so complex?

Comment (nyalldawson) [reply DataPlotly/core/plot_factory.py]: Yes -- because the color mapping used in the color ramp assistant isn't necessarily linear!

Comment (nyalldawson) [reply DataPlotly/core/plot_factory.py]: Maybe we just drop the legend option?

Comment (ghtmtt) [reply DataPlotly/core/plot_factory.py]: Well given that we are having a lot of new features thanks to @SGroe and some other fixes, I'll say that we can drop the legend for the moment and writing an issue as reminder. What is your opinion?

**Comment by ghtmtt** [comment]
@nyalldawson I'm closing this PR for now. Let's keep in mind all the attempts. 

To wrap up: we should find a way to synchronize the QGIS and plotly color ramps creation. Posting here a link to plotly color scales creation: https://plotly.com/python/colorscales/ 

 We should keep in mind:

* continuous vs discrete ramps
* when using a pure numeric variable to color the plot, sometimes it should be treated as a discrete category and not as a numeric one (e.g. a hypothetical `year` variable) 
* plotly has some standard color ramp names (can we *recycle* these names in QGIS or vice versa?)

**Pull request closed without merging**